File: matlab/ch04/unifgam.m
function gam = unifgam(k,R,Q,nu,amn,Nc,x,T) 
% Estimation of activity coefficients using the UNIFAC method 
% input:  
% R,Q: vectors of volumes and surface areas for each functional group 
% nu: number of functional groups contained in each component 
% (row: functional groups k, column: components i) 
% amn: matrix of group interaction parameters 
% Nc: number of components 
% x: vector of liquid-phase mole fraction 
% T: temperature (K) 
% output: 
% gam: vector of activity coefficients for each component 
r = zeros(1,Nc); q = zeros(1,Nc); tau = zeros(k,k); ek = zeros(k,Nc);  
theta = zeros(1,k); beta = zeros(Nc,k); s = zeros(1,k); z = 10;  
for j = 1:Nc, r(j) = sum(R.*nu(:,j)'); q(j) = sum(Q.*nu(:,j)');  
end % row vector(r,q) 
for i = 1:k, ek(i,:) = nu(i,:)*Q(i)./q; end 
tau = exp(-amn/T);  
for i = 1:Nc    
for j = 1:k, beta(i,j) = sum(ek(:,i).*tau(:,j)); end 
end 
for i = 1:k, theta(i) = sum(x.*q.*ek(i,:))/sum(x.*q); end 
for i = 1:k, s(i) = sum(theta'.*tau(:,i)); end 
J = r/sum(r.*x); L = q/sum(q.*x); gamc = 1 - J + log(J) - 5*q.*(1 - J./L + log(J./ L)); gamr = []; 
for i = 1:Nc    
sumb = 0;     
for j = 1:k        
gamval = theta(j)*beta(i,j)/s(j) - ek(j,i)*log(beta(i,j)/s(j));         
sumb = sumb + gamval;    
end     
gamr = [gamr q(i)*(1 - sumb)]; 
end 
gam = exp(gamc + gamr); 
end 

File: python/ch04/unifgam.py
import numpy as np

def unifgam(k, R, Q, nu, amn, Nc, x, T):
    """
    UNIFAC 방법을 사용하여 활동도 계수를 추정합니다.
    
    입력:
    k: 기능기(functional group)의 개수
    R, Q: 각 기능기의 부피 및 표면적 (벡터)
    nu: 각 성분에 포함된 기능기의 수 (행: 기능기 k, 열: 성분 i)
    amn: 그룹 상호작용 매개변수 행렬 (k x k)
    Nc: 성분 수
    x: 액체상 몰 분율 (벡터)
    T: 온도 (K)
    
    출력:
    gam: 각 성분에 대한 활동도 계수 (벡터)
    """
    R = np.array(R)
    Q = np.array(Q)
    nu = np.array(nu)
    amn = np.array(amn)
    x = np.array(x)
    
    # 1. 분자 부피 및 표면적 계산 (r_i, q_i)
    r = np.sum(R[:, np.newaxis] * nu, axis=0)
    q = np.sum(Q[:, np.newaxis] * nu, axis=0)
    
    # 2. 그룹 분율 계산
    # ek: 성분 i 내의 그룹 k의 표면적 분율
    ek = (nu * Q[:, np.newaxis]) / q
    
    # 3. 상호작용 매개변수 (tau)
    tau = np.exp(-amn / T)
    
    # 4. 혼합물 내 그룹 매개변수 (beta)
    # beta[i, j] = sum_k (ek[k, i] * tau[k, j])
    beta = ek.T @ tau
    
    # 5. 혼합물 내 평균 그룹 표면적 분율 (theta) 및 s
    theta = (x * q * ek) / np.sum(x * q)
    theta = np.sum(theta, axis=1) # 각 그룹별 평균
    
    s = theta @ tau
    
    # 6. 조합적(Combinatorial) 기여분 (gamc)
    J = r / np.sum(r * x)
    L = q / np.sum(q * x)
    gamc = 1 - J + np.log(J) - 5 * q * (1 - J / L + np.log(J / L))
    
    # 7. 잔류(Residual) 기여분 (gamr)
    gamr = np.zeros(Nc)
    for i in range(Nc):
        # sumb = sum_j (theta_j * beta_ij / s_j - ek_ji * ln(beta_ij / s_j))
        term1 = (theta * beta[i, :]) / s
        term2 = ek[:, i] * np.log(beta[i, :] / s)
        sumb = np.sum(term1 - term2)
        gamr[i] = q[i] * (1 - sumb)
        
    # 8. 최종 활동도 계수
    gam = np.exp(gamc + gamr)
    
    return gam